src/httpmessage.rs

```diff
@@ -260,6 +261,137 @@ pub trait HttpMessage {
         let boundary = Multipart::boundary(self.headers());
         Multipart::new(boundary, self)
     }
+    
+    /// Return stream of lines.
+    fn readlines(self) -> Readlines<Self>
+    where
+        Self: Stream<Item = Bytes, Error = PayloadError> + Sized,
+    {
+        Readlines::new(self)
+    }
+}
+
+/// Stream to read request line by line.
+pub struct Readlines<T>
+where
+    T: HttpMessage + Stream<Item = Bytes, Error = PayloadError> + 'static,
+{
+    req: T,
+    buff: BytesMut,
+    limit: usize,
+}
+
+impl<T> Readlines<T>
+where
+    T: HttpMessage + Stream<Item = Bytes, Error = PayloadError> + 'static,
+{
+    /// Create a new stream to read request line by line.
+    fn new(req: T) -> Self {
+        Readlines {
+            req,
+            buff: BytesMut::with_capacity(262_144),
+            limit: 262_144,
+        }
+    }
+    
+    /// Change max line size. By default max size is 256Kb
+    pub fn limit(mut self, limit: usize) -> Self {
+        self.limit = limit;
+        self
+    }
+}
+
+impl<T> Stream for Readlines<T>
+where
+    T: HttpMessage + Stream<Item = Bytes, Error = PayloadError> + 'static,
+{
+    type Item = String;
+    type Error = ReadlinesError;
+    
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        let encoding = self.req.encoding()?;
+        // check if there is a newline in the buffer
+        let mut found: Option<usize> = None;
+        for (ind, b) in self.buff.iter().enumerate() {
```

Comment:
Last comment :)

You search buff for every poll even if you already search it before. I’d add attr for indication that buff does not contain new lines.

Reply:
ok. more comments are good actually I have never written something like this before. thank you for your patience.